### PR TITLE
feat: Task execution pipelining

### PR DESCRIPTION
**Pull request by aneojgurhem**
Some things I read to get to this point.

- https://learn.microsoft.com/en-us/dotnet/core/extensions/workers?pivots=dotnet-7-0
- https://learn.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-7.0&tabs=visual-studio
- https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcompletionsource-1?view=net-7.0
- https://learn.microsoft.com/en-us/dotnet/core/extensions/channels

### CODE REVIEW

File: Common/src/Pollster/Pollster.cs [hunk 1]
```diff
@@ -251,40 +270,56 @@ void RecordError(Exception e)
...
+                                              ownerPodName_,
+                                              activitySource_,
+                                              agentHandler_,
+                                              taskHandlerLogger,
+                                              pollsterOptions_,
+                                              () => taskProcessingDict_.Remove(message.TaskId),
+                                              cts);
+
+            taskProcessingDict_.TryAdd(message.TaskId,
+                                       taskHandler);
```
Comment (lemaitre-aneo) [inline Common/src/Pollster/Pollster.cs:319]: check if add was successfull. If id was already in the dictionary, current pod is already processing this very task.

File: Common/src/Pollster/Pollster.cs [hunk 2]
```diff
@@ -293,20 +328,25 @@ await foreach (var message in messages.ConfigureAwait(false))
...
 
                 // If the task was successful, we can remove a failure
                 if (recordedErrors.Count > 0)
                 {
                   recordedErrors.Dequeue();
                 }
               }
+              else
+              {
+                await taskHandler.DisposeAsync()
```
Comment (lemaitre-aneo) [inline Common/src/Pollster/Pollster.cs:347]: This is not called if PreProcessing throws

File: Common/src/Pollster/RunningTaskQueue.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+                               CancellationToken cancellationToken)
+    => await channel_.Writer.WriteAsync(handler,
+                                        cancellationToken)
+                     .ConfigureAwait(false);
+
+  public async Task WaitForNextWriteAsync(TimeSpan          timeout,
+                                          CancellationToken cancellationToken)
+  {
+    var             cts          = new CancellationTokenSource(timeout);
+    await using var registration = cancellationToken.Register(() => cts.Cancel());
```
[the remark this answers is not in the document]
Comment (aneojgurhem) [reply Common/src/Pollster/RunningTaskQueue.cs:49]: I was looking for this one

File: Common/src/Pollster/RunningTaskQueue.cs
[the remark this answers is not in the document]
Comment (aneojgurhem) [reply Common/src/Pollster/RunningTaskQueue.cs:1]: I think is clearer this way

File: Common/src/Pollster/RunningTaskQueue.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+  public bool RemoveException([MaybeNullWhen(false)] out Exception e)
+  {
+    var r = exceptions_.Count > 0;
+
+    e = r
+          ? exceptions_.Dequeue()
+          : null;
+
+    return r;
+  }
```
[the remark this answers is not in the document]
Comment (aneojgurhem) [reply Common/src/Pollster/RunningTaskQueue.cs:77]: Not as clear and easy to use.

**Comment by aneojgurhem** [comment]
I did some tests on AWS with a c5.12large instance.

install dependencies

```
sudo yum install -y git docker yum-utils

# just
mkdir -p ~/bin
curl --proto '=https' --tlsv1.2 -sSf https://just.systems/install.sh | bash -s -- --to ~/bin

# terraform
sudo yum-config-manager --add-repo https://rpm.releases.hashicorp.com/AmazonLinux/hashicorp.repo
sudo yum -y install terraform

# docker
sudo systemctl enable docker.service
sudo systemctl start docker.service
sudo usermod -a -G docker ec2-user
```

Relog

```
just worker=mock replicas=30 partitions=1 build-deploy
```

**Review by lemaitre-aneo — APPROVED** [review]

**Review by lemaitre-aneo — APPROVED** [review]

**Pull request merged**